src/server.rs

```diff
@@ -271,35 +269,50 @@ impl ServerPlugin {
     ) -> bincode::Result<()> {
         replicated_archetypes.update(set.p0(), &rules);
 
-        let replicated_clients = mem::take(&mut *set.p1()); // Take ownership to avoid borrowing issues.
-        messages.prepare(replicated_clients);
+        // Take ownership to avoid borrowing issues.
+        let mut replicated_clients = mem::take(&mut *set.p1());
+        let mut removal_buffer = mem::take(&mut *set.p2());
+        let mut client_buffers = mem::take(&mut *set.p3());
 
-        collect_mappings(&mut messages, &mut set.p2())?;
-        collect_despawns(&mut messages, &mut set.p3())?;
-        collect_removals(&mut messages, &mut set.p4(), &mut entities_with_removals)?;
+        messages.reset(replicated_clients.len());
+
+        collect_mappings(&mut messages, &replicated_clients, &mut set.p4())?;
+        collect_despawns(&mut messages, &mut replicated_clients, &mut set.p5())?;
+        collect_removals(&mut messages, &removal_buffer)?;
         collect_changes(
             &mut messages,
+            &mut replicated_clients,
             &replicated_archetypes,
             &registry,
-            &entities_with_removals,
+            &removal_buffer,
             set.p0(),
             &change_tick,
             **server_tick,
         )?;
-        entities_with_removals.clear();
-
-        let mut client_buffers = mem::take(&mut *set.p5());
-        let replicated_clients = messages.send(
-            &mut set.p6(),
-            &mut client_buffers,
-            **server_tick,
-            change_tick.this_run(),
-            time.elapsed(),
-        )?;
+        removal_buffer.clear();
+
+        for ((init_message, update_message), client) in
```

Comment:
```suggestion
        // Sends messages to clients.
        //
        // The change tick of each client with an init message is updated to equal the latest replicon tick
        // messages were sent to clients. If only update messages were sent (or no messages at all) then
        // it will equal the `last_change_tick`.
        for ((init_message, update_message), client) in
```

Reply:
Will restore!

Reply:
This is a higher-level comment that describes the overall effect of sending init and update messages. I think we should keep it around.